enhancements/api-review/custom-tags-aws.md

```diff
@@ -31,113 +35,294 @@ have the ability to apply user defined tags to many resources created by OpenShi
+  Action = Existing tag for AWS resource is updated to reflect new value.
+  Final tag set to AWS resource = `key_infra1 = value_update1`
+  Event action = An event is generated to notify user about the action status (success/failure) to update tags for the AWS resource. A warning also should be generated that value is set for deletion.
+
+  Second pass,
+  Existing tag for AWS resource = `key_infra1 =`
+  New tag request = `.spec.platformSpec.aws.resourceTags` has `key_infra1 =`
+  Action = Existing tag for AWS resource is deleted.
+  Final tag set to AWS resource = deleted
+  Event action = An event is generated to notify user about the action status (success/failure) to delete tags for the AWS resource.
+
+### Caveats
+1) User updates the user-defined tag from using external tools when there is an entry in `.spec.platformSpec.aws.resourceTags`
+   The user-defined tag which is updated from spec, will be reconciled by operators to set value from `.spec.platformSpec.aws.resourceTags`.
+   The user-defined tag will be overwritten with value from `.spec.platformSpec.aws.resourceTags` when there is an update to `.spec.platformSpec.aws.resourceTags` section as a whole.
+
+   User must handle inconsistencies in `.spec.platformSpec.aws.resourceTags` and user-defined tag value for AWS resource when using multiple tools to manage tags.
+
+   For example,
+   Edited existing tag using external tool for AWS resource = `key_infra1 = value_comp1`
+   Previous tag request = `.spec.platformSpec.aws.resourceTags` has `key_infra1 = value_infra1`
+   Action = Update existing tag with value from `.spec.platformSpec.aws.resourceTags`.
+   Final tag set to AWS resource = `key_infra1 = value_infra1`
+   Event action = An event is generated to notify user about the action status (success/failure) to update tags for the AWS resource.
+
+2) User deletes the user-defined tag from `.spec.platformSpec.aws.resourceTags`
+   The user-defined tag which is removed from spec, will not be reconciled or managed by operators.
+   User can update user-defined tag key:value using external tool. The user-defined tag will not be overwritten.
+
+   For example,
+   Existing tag for AWS resource = `key_infra1 = value1`
+   New tag request = `.spec.platformSpec.aws.resourceTags` has no user-defined tag with key `key_infra1`
+   Action = No change in existing user-defined tag.
+   Final tag set to AWS resource = `key_infra1 = value1`
+   Event action = No event
+
+3) User sets user-defined tag to delete which is created using external tools.
+   There is no validation check involved for creator of user-defined tag. Any user-defined tag added by user in `.spec.platformSpec.aws.resourceTags` is considered for create/update/delete accordingly.
+
+4) Any user-defined tag set using `.spec.platformSpec.aws.resourceTags` in `Infrastructure.config.openshift.io/v1` type has scope limited to cluster-level. Individual AWS resource-level user-defined tag management is not supported.
```

Comment:
During creation, user tags are read from install config and applied to all resources created for the cluster. Update is handled by the operators which manage these resources. For example, in case of machine scale operations, tags will be applied as per the values in spec from Infrastructure CRD.

Reply:
> For example, in case of machine scale operations, tags will be applied as per the values in spec from Infrastructure CRD.

That's not the behaviour that we exhibit today and making that true would be a breaking change.

If a user creates a MachineSet at a later date, what happens about the tags? They don't have the infra tags as it wasn't created during install, they can put whatever they like on there, there's nothing to stop them putting a value which conflicts with the Infrastructure resource, and IMO, they should be able to. The Machine spec is closest to the creation of the resource and should be highest precedence unless absolutely critical not to (eg the Name or cluster ID tags).

I don't think we can change this behaviour between releases. What are controllers outside of Machine API doing? How many different CRDs in OpenShift allow tags at the moment?